### PR TITLE
[SofaBaseLinearSolver] FIX breaking condition in CG at first step regarding threshold

### PR DESCRIPTION
**Pull request by hugtalbot**
Fix bug introduced in #521 

If threshold is met at first step with den=0, break but info msg
about the status of the linear system.

If threshold is met at first with den!=0, then continue one more iteration
and warning about threshold value too large.

______________________________________________________

This PR: 
- [x] builds with SUCCESS for all platforms on the CI.
- [x] does not generate new warnings.
- [x] does not generate new unit test failures.
- [x] does not generate new scene test failures.
- [x] does not break API compatibility.
- [x] is more than 1 week old (or has fast-merge label).

**Reviewers will merge only if all these checks are true.**

### CODE REVIEW

**Review by IPeterlik — COMMENTED** [review]
Looks OK to me. Basically, apart from issuing a warning, the code should not explicitly handle the case when the residual is below a threshold already in the first iteration, since numerically, it's a correct behavior even if it yields an unwanted result (e.g., an object is not moving despite applied forces). CG is an iterative solver and anyone using it should be well aware of its parameters and their impact on the simulation.

**Comment by damienmarchal** [comment]
[ci-build][with-scene-tests]

File: SofaKernel/modules/SofaBaseLinearSolver/CGLinearSolver.inl [hunk 1]
```diff
@@ -195,54 +227,60 @@ void CGLinearSolver<TMatrix,TVector>::solve(Matrix& M, Vector& x, Vector& b)
...
+            {
+                msg_warning() << "denominator threshold reached at first iteration of CG" << msgendl
+                              << "Check the 'threshold' data field, you might decrease it";
+            }
+
             endcond = "threshold";
             if( verbose )
             {
-                sout<<"CGLinearSolver, den = "<<den<<", smallDenominatorThreshold = "<<f_smallDenominatorThreshold.getValue()<<sendl;
+                msg_info() << "CGLinearSolver, den = " << den <<", smallDenominatorThreshold = " << f_smallDenominatorThreshold.getValue();
```
Comment (damienmarchal) [inline SofaKernel/modules/SofaBaseLinearSolver/CGLinearSolver.inl]: No need for CGLinearSolver...msg_info add the prefix.

File: SofaKernel/modules/SofaBaseLinearSolver/CGLinearSolver.inl [hunk 2]
```diff
@@ -195,54 +227,60 @@ void CGLinearSolver<TMatrix,TVector>::solve(Matrix& M, Vector& x, Vector& b)
...
         /// End of the CG step : update x and r
         cgstep_alpha(params, x,r,p,q,alpha);
 
-
         if( verbose )
         {
-            sout<<"den = "<<den<<", alpha = "<<alpha<<sendl;
-            sout<<"x : "<<x<<sendl;
-            sout<<"r : "<<r<<sendl;
+            msg_info() << "den = " << den << ", alpha = " << alpha;
```
Comment (damienmarchal) [inline SofaKernel/modules/SofaBaseLinearSolver/CGLinearSolver.inl]: No needs for three messages...one is better (and more efficient)

File: SofaKernel/modules/SofaBaseLinearSolver/CGLinearSolver.inl [hunk 3]
```diff
@@ -63,6 +63,29 @@ CGLinearSolver<TMatrix,TVector>::CGLinearSolver()
     f_smallDenominatorThreshold.setRequired(true);
 }
 
+template<class TMatrix, class TVector>
+void CGLinearSolver<TMatrix,TVector>::init()
```
Comment (damienmarchal) [inline SofaKernel/modules/SofaBaseLinearSolver/CGLinearSolver.inl:67]: For usability reason I would recommand to have something like
```cpp
if(f_verbose.getValue()){
    f_printLog.setValue(true); 
}

File: SofaKernel/modules/SofaBaseLinearSolver/CGLinearSolver.inl [hunk 4]
```diff
@@ -109,7 +132,7 @@ void CGLinearSolver<TMatrix,TVector>::solve(Matrix& M, Vector& x, Vector& b)
     double rho, rho_1=0, alpha, beta;
 
 
-    msg_info_when(verbose) <<" CGLinearSolver, b = "<< b ;
+    msg_info_when(verbose) << " CGLinearSolver, b = " << b ;
```
Comment (damienmarchal) [inline SofaKernel/modules/SofaBaseLinearSolver/CGLinearSolver.inl]: No needs for the 'CGLinearSolver' prefix

**Comment by damienmarchal** [comment]
Now it is fast merge :)

**Comment by damienmarchal** [comment]
For me, when the CI will re-vive and says ok it may go.

**Comment by bcarrez** [comment]
[ci-build]

**Comment by damienmarchal** [comment]
[ci-build][with-scene-tests]

**Comment by damienmarchal** [comment]
[ci-build][with-scene-tests]

**Comment by damienmarchal** [comment]
[ci-build][with-scene-tests]

**Comment by damienmarchal** [comment]
The dashboard is broken...I bet it is because of the "weird" branch name :)

**Comment by damienmarchal** [comment]
The dashboard is broken...I bet it is because of the "weird" branch name :)
[ci-build][with-scene-tests]

**Comment by damienmarchal** [comment]
[ci-build][with-scene-tests]

**Comment by damienmarchal** [comment]
[ci-build][with-scene-tests]

**Comment by damienmarchal** [comment]
[ci-build][with-scene-tests]

**Comment by damienmarchal** [comment]
[ci-build][with-scene-tests]

**Comment by damienmarchal** [comment]
Ok the PR' works, I looked at each build on jenkins...no failure...but the formatter is not launched but I think we can ignore the formatting stuff to move forward the release. @guparan @bcarrez do you agree ?

PS: the sofa dashboard doesn't like PR with weird name & symbols :)

**Comment by guparan** [comment]
Agreed, let's merge!

**Pull request merged**